receiver/zookeeperreceiver/scraper.go

```diff
@@ -16,31 +16,134 @@ package zookeeperreceiver
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"net"
+	"regexp"
+	"strconv"
+	"time"
 
 	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/simple"
 	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/zookeeperreceiver/internal/metadata"
+)
+
+var zookeeperFormatRE = regexp.MustCompile(`(^zk_\w+)\s+([\w\.\-]+)`)
+
+const (
+	mntrCommand = "mntr"
 )
 
 type zookeeperMetricsScraper struct {
-	cfg *Config
+	logger *zap.Logger
+	config *Config
+	cancel context.CancelFunc
 }
 
-func newZookeeperMetricsScraper(_ *zap.Logger, cfg *Config) (*zookeeperMetricsScraper, error) {
-	return &zookeeperMetricsScraper{cfg: cfg}, nil
+func (z *zookeeperMetricsScraper) Name() string {
+	return typeStr
 }
 
-func (z *zookeeperMetricsScraper) Name() string {
-	return z.cfg.Name()
+func newZookeeperMetricsScraper(logger *zap.Logger, config *Config) (*zookeeperMetricsScraper, error) {
+	_, _, err := net.SplitHostPort(config.TCPAddr.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	if config.Timeout <= 0 {
+		return nil, errors.New("timeout must be a positive duration")
+	}
+
+	return &zookeeperMetricsScraper{
+		logger: logger,
+		config: config,
+	}, nil
 }
 
 func (z *zookeeperMetricsScraper) Initialize(_ context.Context) error {
 	return nil
 }
 
 func (z *zookeeperMetricsScraper) Close(_ context.Context) error {
+	z.cancel()
 	return nil
 }
 
-func (z *zookeeperMetricsScraper) Scrape(_ context.Context, _ string) (pdata.ResourceMetricsSlice, error) {
-	return pdata.ResourceMetricsSlice{}, nil
+func (z *zookeeperMetricsScraper) Scrape(ctx context.Context, _ string) (pdata.ResourceMetricsSlice, error) {
+	var ctxWithTimeout context.Context
+	ctxWithTimeout, z.cancel = context.WithTimeout(ctx, z.config.Timeout)
+
+	conn, err := z.config.Dial()
+	if err != nil {
+		z.logger.Error("failed to establish connection",
+			zap.String("endpoint", z.config.Endpoint),
+			zap.Error(err),
+		)
+		return pdata.ResourceMetricsSlice{}, err
+	}
+	defer conn.Close()
+
+	deadline, ok := ctxWithTimeout.Deadline()
+	if ok {
+		conn.SetDeadline(deadline)
+	}
+
+	return z.getResourceMetrics(conn), nil
+}
+
+func (z *zookeeperMetricsScraper) getResourceMetrics(conn net.Conn) pdata.ResourceMetricsSlice {
+	scanner := sendCmd(conn, mntrCommand)
```

Comment:
Since the connection and the `sendCmd` function are so closely related, an option would be to wrap the connection in a type and have `sendCmd` be a method on that type. More generally, I think the I/O in this file/type could be moved into that type, so that the scraper is 'skinny'. Just a thought though -- up to you.

Reply:
I played around a bit with this approach, however, I wasn't able figure out a way to cleanly mock some of the methods for testing error cases. Having some of these methods on the scraper struct makes it easier to mock in tests IMO.

Reply:
Ok, no problem. We can always do this in a separate PR if we change our minds.